[Core] Enable prefix caching with block manager v2 enabled

leiwen83 opened the pull request:
[Core] Enable prefix caching with block manager v2 enabled

    This PR is for enabel APC feature over block manager V2.

    APC core idea is like maintain a table mapping between content hash and
    physical block. If one block is full, then it would be prompted to be
    immutable. While if several immutable comes in sequence, we kept the
    first physical block, while free the other.

    Those freed physical block, aka no reference at scheduler, would be
    housekept inside evictor, when we try to alloc block again under pressue,
    and find no free one as in the initial state, we could resort to
    evictor to pop one unrefered computed block, which is used to store
    new data comes from upper layer.

    The evictor policy could be changed, but current only implement LRU one.

    Co-authored-by: Sage Moore <sagemoore@utexas.edu>

leiwen83 commented:
@cadedaniel 
Would you mind take a look for this patch to enable the prefix caching with block manager V2?

SageMoore commented:
Hi, @leiwen83 looks like we collided a bit. I just posted https://github.com/vllm-project/vllm/pull/4146 which also adds prefix caching support to V2 of the block manager. Would you be willing to work together to get that diff over the finish line? The main thing I like from this PR is that it stores references to the blocks in the PrefixCachingBlockAllocator. I did something similar with a new BlockMetaData structure, but I think we can make some minor adjustments to the existing LRU evictor such that it will work with generic block objects. That way neither one of us has to add a new, but basically the same, evictor :). 

Occasional duplication of work is an unfortunate reality with popular open source projects, but I think we can work together and end up merging something pretty good.

leiwen83 commented:
Hi @SageMoore 

what a coincidence ;)

But I'd like to hear from @cadedaniel first before take further action, he may give us some good advice.

Thx,
Lei

cadedaniel commented:
Thanks @leiwen83 and @SageMoore for the work ~~

Priority-wise, whoever has a more complete reviewable PR should get review precedence. I'll look at this one first as it's no longer WIP / tests are passing. @SageMoore feel free to add review comments on missing requirements or improvements. I haven't thought through the Evictor design tradeoffs so especially clarifying which approach we should go with for that would be helpful!

In the future we should be more communicative in https://github.com/vllm-project/vllm/issues/3667 so we don't have mixups

SageMoore reviewed (changes requested):
I think moving incrementally here is good. Let's try to get this PR into a mergable state and go from there. 

There are four high level changes that I would suggest for this PR.
1. Move the core/block/evictor.py file to core/evictor_v2.py, similar to what was done for block_manager. Rename core/evictor.py to core/evictor_v1.py.
2. Let's remove unused_cached_blocks from the PrefixCachingBlockAllocator and just rely on the evictor.
3. Assuming that we don't want to make the policy an argument to the BlockSpaceManagerV2, lets move the code that records last_accessed and computed into the PrefixCachingBlockAllocator. The code to manage and record these two variables will likely change if a non-LRU policy is used. For example, we wouldn't want to spend time recording last_accessed time if we switched to a FIFO eviction policy. 
4. Evictor unit tests would be good. The eviction logic is somewhat cumbersome to exercise in integration tests because it only runs when we run out of blocks.

SageMoore commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -44,6 +47,9 @@ def __init__(
         self._unused_cached_blocks: Dict[PrefixHash, BlockId] = {}
 
+        # A mapping of blockId to Block to track those cached blocks
+        self._blocks: Dict[BlockId, Block] = {}
```
Is the intention to have this track all blocks allocated by the _hashless_allocator or just the blocks in _cached_blocks?

leiwen83 replied:
It server for track mapping in the _cached_blocks, as mark_blocks_as_computed/get_common_computed_block_ids's func param past from block manager is block_id only, and it has no idea for underlying content hash

SageMoore commented on vllm/core/block/evictor.py:
```diff
@@ -0,0 +1,110 @@
+        pass
+
+
+class Block():
```
There's a bit of a name collision here. Can you rename this since "Block" is already used extensively? I went with BlockMetaData. Feel free to use that or a new name :). EvictorBlockInfo?

leiwen83 replied:
BlockMetaData is nice, I'll take it~

SageMoore commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -351,6 +376,9 @@ def __init__(
         self._cached_content_hash: Optional[int] = None
         self._prefix_caching_allocator = prefix_caching_allocator
+        self.computed = False
+        self.num_hashed_tokens = 0
```
Nit: Instead of manually setting this when marking blocks as computed, you should be able to do something like:
```
    @property
    def num_tokens_total(self) -> int:
        return self._prev_block.num_tokens_total + len(self.token_ids)
```
You'll have to account for the _prev_block == None case.

leiwen83 replied:
I didn't get the num_tokens_total usage here...

leiwen83 commented:
> I think moving incrementally here is good. Let's try to get this PR into a mergable state and go from there.
> 
> There are four high level changes that I would suggest for this PR.
> 
> 1. Move the core/block/evictor.py file to core/evictor_v2.py, similar to what was done for block_manager. Rename core/evictor.py to core/evictor_v1.py.
> 2. Let's remove unused_cached_blocks from the PrefixCachingBlockAllocator and just rely on the evictor.
> 3. Assuming that we don't want to make the policy an argument to the BlockSpaceManagerV2, lets move the code that records last_accessed and computed into the PrefixCachingBlockAllocator. The code to manage and record these two variables will likely change if a non-LRU policy is used. For example, we wouldn't want to spend time recording last_accessed time if we switched to a FIFO eviction policy.
> 4. Evictor unit tests would be good. The eviction logic is somewhat cumbersome to exercise in integration tests because it only runs when we run out of blocks.


Great thanks for those comments! And I borrow the evictor test case from your PR, and add some change to make it pass.
Let's see what else need to be done for this PR.

SageMoore commented:
This looks reasonable to me. Thanks for working on it!

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -133,20 +133,21 @@ def allocate_mutable(self, prev_block: Block) -> Block:
 
             # Clear content hash mapping; the block will be overwritten.
-            del self._cached_blocks[content_hash_to_evict]
+            del self._blocks[self._cached_blocks[content_hash_to_evict]]
+            block_id = self._cached_blocks.pop(content_hash_to_evict)
```
nit: reverse the order to require only 1 hash lookup instead of 2

cadedaniel replied:
btw we will need to profile the BlockManagerV2. my understanding from https://github.com/vllm-project/vllm/pull/4270 is that the scheduler is pretty sensitive to µs overhead

rkooo567 replied:
yeah if the function is called for each iteration https://github.com/vllm-project/vllm/blob/c1b4e4157c0b4154f950adaea85a259fc629c758/vllm/core/scheduler.py#L429, having additional 5+us overhead for the function slows down the performance

leiwen83 replied:
Got it, it make sense, would apply this change

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -7,9 +7,11 @@
+DEFAULT_LAST_ACCESSED_TIME = -1
 
 
 class PrefixCachingBlockAllocator(BlockAllocator):
```
can we update docstring with your improvements?

leiwen83 replied:
Ok, I would put a docstring here to illustract what this varaible is for

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -54,6 +53,8 @@ def __init__(
 
         self._block_size = block_size
 
+        self.evictor: Evictor = make_evictor(eviction_policy)
```
nit: add some comment explaining the role

leiwen83 replied:
Got it~

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -157,6 +158,8 @@ def allocate_mutable(self, prev_block: Block) -> Block:
                 block_id=block_id,
             )
             assert block.content_hash is None
+            block.computed = False
```
is this necessary? the prefix caching block constructor will set it to False

leiwen83 replied:
I see, would remove it

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -133,20 +133,21 @@ def allocate_mutable(self, prev_block: Block) -> Block:
-        if self._unused_cached_blocks:
-            # TODO policy for selecting block to remove
-            content_hash_to_evict = next(iter(self._unused_cached_blocks))
+        if self.evictor.num_blocks > 0:
```
nit: add a comment explaining this branch
```
# If the evictor has blocks available for eviction, evict a block and return it.
```

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -191,11 +194,11 @@ def _free_block_id_for_block(self, block_id: BlockId,
-            assert block.content_hash not in self._unused_cached_blocks
             assert block.content_hash in self._cached_blocks
-            self._unused_cached_blocks[block.content_hash] = block_id
+            self.evictor.add(block.content_hash, block.block_id,
+                             block.num_tokens_total, block.last_accessed)
```
What happens if a block is not computed yet?

leiwen83 replied:
there is "assert block.content_hash in self._cached_blocks" to make sure this block shall have content_hash already. So we may safely claim it is a computed block?

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -231,8 +234,8 @@ def fork(self, last_block: Block) -> List[Block]:
     def get_num_free_blocks(self) -> int:
         # The number of free blocks is the number of hashless free blocks
         # plus the number of hashful blocks that are unused.
```
nit: update comment

leiwen83 replied:
got it

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -293,10 +296,28 @@ def clear_copy_on_writes(self) -> Dict[BlockId, List[BlockId]]:
+        Only mark those immutable block as computed.
+        """
+        for block_ids in seq_block_ids:
+            block_cnt = 0
```
why do we need to track `block_cnt`?

leiwen83 replied:
This is no needed and forget to delete, which served for num_token computation

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -293,10 +296,28 @@ def clear_copy_on_writes(self) -> Dict[BlockId, List[BlockId]]:
+    def mark_blocks_as_computed(self, seq_block_ids: List[List[int]]) -> None:
+        """Mark immutable blocks as computed, used in prefix caching.
+
+        Only mark those immutable block as computed.
```
how does this check if a block is immutable? seems it doesn't

leiwen83 replied:
Another thought for this mark_blocks_as_computed is whether we could deem those blocks already has content_hash as computed one?

If so, mark_blocks_as_computed could be made as empty. To check it is computed, just need a property to check whether it has a valid content_hash

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -305,15 +326,14 @@ def get_common_computed_block_ids(
         ids_list = [
-            takewhile(lambda block_id: computed(block_id), seq[:-1])
-            for seq in seq_block_ids
+            list(
+                takewhile(lambda block_id: self._blocks[block_id].computed,
+                          seq[:-1])) for seq in seq_block_ids
```
This should have a unit test. cc @SageMoore .

I won't block the PR on it because you didn't add this code yourself, but it would be great to add.

leiwen83 replied:
I see, would add a test case for it

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -7,9 +7,11 @@
 
 PrefixHash = int
 BlockId = int
+DEFAULT_LAST_ACCESSED_TIME = -1
```
nit: `_DEFAULT_LAST_ACCESSED_TIME`

leiwen83 replied:
Got it, would apply~

cadedaniel commented on vllm/core/block_manager_v2.py:
```diff
@@ -72,14 +72,13 @@ def __init__(
         self.watermark_blocks = int(watermark * num_gpu_blocks)
 
         self.block_allocator = CpuGpuBlockAllocator.create(
             # Currently, only naive blocks are supported (no prefix caching).
```
remove comment

leiwen83 replied:
Got it, would apply~

cadedaniel commented on vllm/core/block_manager_v2.py:
```diff
@@ -194,17 +193,23 @@ def get_block_table(self, seq: Sequence) -> List[int]:
+            # in this step.
+            block_table = self.block_tables[seq.seq_id]
+            block_ids = []
+            for block in block_table._blocks:
```
we should avoid using protected APIs, e.g. `_blocks`. can we use this instead?
https://github.com/vllm-project/vllm/blob/ceaf4ed0030c7816537359b8efc750474149ce0f/vllm/core/block/block_table.py#L182

leiwen83 replied:
Got it, would apply~

cadedaniel commented on vllm/core/block_manager_v2.py:
```diff
@@ -194,17 +193,23 @@ def get_block_table(self, seq: Sequence) -> List[int]:
-        # Tracked here https://github.com/vllm-project/vllm/issues/3667
-        pass
+    def access_all_blocks_in_seq(self, seq: Sequence, now: float):
+        if self.enable_caching:
```
nit: add comment why we only do this for prefix caching (seems performance?)

leiwen83 replied:
Got it, would apply~

cadedaniel commented on vllm/core/block/cpu_gpu_block_allocator.py:
```diff
@@ -190,10 +190,16 @@ def clear_copy_on_writes(self) -> Dict[int, List[int]]:
+    def mark_blocks_as_computed(self, seq_block_ids: List[int]) -> None:
+        # Prefix caching only supported on GPU.
+        device = Device.GPU
+        return self._allocators[device].mark_blocks_as_computed(seq_block_ids)
```
can we validate that the block ids are on GPU? otherwise the error message will be pretty confusing

leiwen83 replied:
I tend to the logic of check whether block has content_hash for judging whether it is computed, so we no longer need this function in cpu_gpu_block_allocator.

cadedaniel commented on vllm/core/block/cpu_gpu_block_allocator.py:
```diff
@@ -190,10 +190,16 @@ def clear_copy_on_writes(self) -> Dict[int, List[int]]:
         # Prefix caching only supported on GPU.
         device = Device.GPU
-        return self._allocators[device].mark_blocks_as_computed()
+        return self._allocators[device].mark_blocks_as_accessed(block_ids, now)
```
can we validate that the block ids are on GPU? otherwise the error message will be pretty confusing

leiwen83 replied:
In prefix_caching_block.py, I track the already cached blocks in _blocks. only mark those id if they are in the cached pool, I think it shall be enough for validation of GPU.

leiwen83 replied:
I see, I would raise error when found the block is not belong to gpu vllm/core/block/prefix_caching_block.py.

cadedaniel reviewed (commented):
Thanks for the PR. Some missing things:
* Can you add a PR description? it helps with fast reviews -- you can detail your approach, the testing, what's missing, etc. Shouldn't take long!
* To call automatic prefix caching in BlockManagerV2 "done", we need end-to-end tests.
  * See this example https://github.com/vllm-project/vllm/blob/ceaf4ed0030c7816537359b8efc750474149ce0f/tests/core/block/e2e/test_correctness.py#L187-L199, which compares decoding output of a sequence with lookahead scheduling against output without lookahead scheduling and verifies they are the same.
  * Off the top of my head, we'll want to test the following in APC. Maybe @SageMoore can also help here, since it's a bit out of scope of the BlockManagerV2 itself.
    * Greedy output equality with APC and without APC
    * Greedy output equality with APC + preemption

leiwen83 commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -398,6 +420,13 @@ def is_full(self) -> bool:
     def num_empty_slots(self) -> int:
         return self._block.num_empty_slots
 
+    @property
+    def num_tokens_total(self) -> int:
+        if self._prev_block is None:
+            return len(self.token_ids)
+
+        return self._prev_block.num_tokens_total + len(self.token_ids)
```
One throught here is to have a variable cache the compute result, so O(seq_len) is only for the first time, then later shall have O(1).
However this need an assumption that the chain won't break, so that previous num_token cached still hold valid, I'm not sure whether it is true in current V2 design?

leiwen83 replied:
Got it, would apply similar design here.

leiwen83 commented on vllm/core/block/cpu_gpu_block_allocator.py:
```diff
@@ -190,10 +190,16 @@ def clear_copy_on_writes(self) -> Dict[int, List[int]]:
-        return self._allocators[device].mark_blocks_as_computed()
+        return self._allocators[device].mark_blocks_as_accessed(block_ids, now)
+
+    def mark_blocks_as_computed(self, seq_block_ids: List[int]) -> None:
```
mark_blocks_as_computed now a null operation for block v2

leiwen83 commented on vllm/core/block/cpu_gpu_block_allocator.py:
```diff
@@ -190,10 +190,16 @@ def clear_copy_on_writes(self) -> Dict[int, List[int]]:
         return self._allocators[device].clear_copy_on_writes()
 
-    def mark_blocks_as_computed(self) -> None:
+    def mark_blocks_as_accessed(self, block_ids: List[int],
```
Got it, would apply~

wenlei03 committed:
[Core] Enable prefix caching with block manager v2 enabled

This PR is for enabel APC feature over block manager V2.

APC core idea is like maintain a table mapping between content hash and
physical block. If one block is full, then it would be prompted to be
immutable. While if several immutable comes in sequence, we kept the
first physical block, while free the other.

Those freed physical block, aka no reference at scheduler, would be
housekept inside evictor, when we try to alloc block again under pressue,
and find no free one as in the initial state, we could resort to
evictor to pop one unrefered computed block, which is used to store
new data comes from upper layer.

The evictor policy could be changed, but current only implement LRU one.

Co-authored-by: Sage Moore <sagemoore@utexas.edu>

leiwen83 commented:
> Thanks for the PR. Some missing things:
> 
> * Can you add a PR description? it helps with fast reviews -- you can detail your approach, the testing, what's missing, etc. Shouldn't take long!
> * To call automatic prefix caching in BlockManagerV2 "done", we need end-to-end tests.
>   
>   * See this example https://github.com/vllm-project/vllm/blob/ceaf4ed0030c7816537359b8efc750474149ce0f/tests/core/block/e2e/test_correctness.py#L187-L199
>     , which compares decoding output of a sequence with lookahead scheduling against output without lookahead scheduling and verifies they are the same.
>   * Off the top of my head, we'll want to test the following in APC. Maybe @SageMoore can also help here, since it's a bit out of scope of the BlockManagerV2 itself.
>     
>     * Greedy output equality with APC and without APC
>     * Greedy output equality with APC + preemption

Got it, I refactor the PR and make the change as suggested. For e2e test, I add two, one is prefill caching enabled while comparing v1 vs v2, and in v2 + preemption, w/ APC vs w/o APC.

wenlei03 committed:
fix ruff

leiwen83 commented on tests/core/block/test_prefix_caching_block.py:
```diff
@@ -358,6 +358,181 @@ def test_get_num_free_blocks_shared(num_blocks: int, block_size: int,
+                                       seed: int):
+        """Verify sharing occurs by allocating two sequences that share prefixes
+        and incrementally freeing blocks.
+        """
```
Got it, refresh that~

leiwen83 commented on tests/core/block/test_prefix_caching_block.py:
```diff
@@ -358,6 +358,181 @@ def test_get_num_free_blocks_shared(num_blocks: int, block_size: int,
+    @pytest.mark.parametrize("seed", list(range(20)))
+    def test_mark_and_get_computed(num_blocks: int, block_size: int,
+                                   seed: int):
+        """Verify sharing occurs by allocating two sequences that share prefixes
```
Got it, refresh that~

leiwen83 commented on tests/core/block/test_prefix_caching_block.py:
```diff
@@ -358,6 +358,181 @@ def test_get_num_free_blocks_shared(num_blocks: int, block_size: int,
+    @pytest.mark.parametrize("seed", list(range(20)))
+    def test_eviction_order(num_blocks: int, block_size: int, seed: int):
+        """Verify sharing occurs by allocating two sequences that share prefixes
+        and incrementally freeing blocks.
```
Got that, refresh that~

leiwen83 commented on tests/core/block/test_prefix_caching_block.py:
```diff
@@ -358,6 +358,181 @@ def test_get_num_free_blocks_shared(num_blocks: int, block_size: int,
+        ]
+        second_computed_ids = [
+            second_chain[i].block_id for i in range(num_blocks_to_consume // 4)
+        ]
```
I change this test case a bit, and now it shall looks better.

wenlei03 committed:
fix yapf

wenlei03 committed:
fix _blocks track issue

wenlei03 committed:
fix ruff issue

wenlei03 committed:
fix yapf

cadedaniel commented on vllm/core/block/prefix_caching_block.py:
```diff
@@ -398,6 +473,24 @@ def is_full(self) -> bool:
+        if self._cached_num_tokens_total is not None:
+            return self._cached_num_tokens_total
+
+        _block = self
+        self._cached_num_tokens_total = 0
+        while _block is not None:
+            self._cached_num_tokens_total += len(_block.token_ids)
+            _block = _block.prev_block
```
This is still O(N^2) for the entire sequence, e.g. each block will perform O(N) scan. We should use the design I linked to you [`content_hash`](https://github.com/vllm-project/vllm/blob/4bb53e2dde809ea5727b8cac95a080893733a1ef/vllm/core/block/prefix_caching_block.py#L414) which is O(N).

I will not block the PR (we can optimize later) but you should add a TODO!

leiwen83 replied:
I see, I would add this TODO note

cadedaniel reviewed (commented):
Please add a TODO on the O(N^2) num_total_tokens. Otherwise LGTM.

wenlei03 committed:
add performance todo note

cadedaniel reviewed (approved):
Thanks @leiwen83
